src/ai/backend/common/events/dispatcher.py

```diff
@@ -0,0 +1,527 @@
+from __future__ import annotations
+
+import asyncio
+import enum
+import logging
+import secrets
+import time
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from typing import (
+    Any,
+    Callable,
+    Coroutine,
+    Generic,
+    Optional,
+    Protocol,
+    Self,
+    Type,
+    TypedDict,
+    TypeVar,
+    Union,
+    cast,
+)
+
+import attrs
+from aiomonitor.task import preserve_termination_log
+from aiotools.taskgroup import PersistentTaskGroup
+from aiotools.taskgroup.types import AsyncExceptionHandler
+
+from ai.backend.common.events.user_event.user_event import UserEvent
+from ai.backend.common.message_queue.queue import AbstractMessageQueue
+from ai.backend.logging import BraceStyleAdapter
+
+from .. import msgpack
+from ..types import (
+    AgentId,
+)
+
+__all__ = (
+    "AbstractEvent",
+    "EventCallback",
+    "EventDispatcher",
+    "EventHandler",
+    "EventProducer",
+)
+
+log = BraceStyleAdapter(logging.getLogger(__spec__.name))
+
+
+class EventDomain(enum.StrEnum):
+    BGTASK = "bgtask"
+    IMAGE = "image"
+    KERNEL = "kernel"
+    MODEL_SERVICE = "model_service"
+    ROUTE = "route"
```

Comment:
Does the Route need to be a separate domain from the Model Service?
Even if they need to be separate, there seems to be a more intuitive name than "ROUTE."

Reply:
I'll combine them into model serving.